### PR TITLE
feat(storage): update  interpreter and storage support

### LINKED ISSUE #7349: feat: Update
**Summary**

After `delete` supported, it's time for `update`.

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

```
mysql> create table t(a int, b int);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values(2,2);
Query OK, 1 row affected (0.01 sec)

mysql> insert into t values(1,1);
Query OK, 1 row affected (0.02 sec)

mysql> update t set a=2 where b=1;
Query OK, 1 row affected (0.03 sec)

mysql> select * from t;
+------+------+
| a    | b    |
+------+------+
|    2 |    1 |
|    2 |    2 |
+------+------+
2 rows in set (0.01 sec)
Read 2 rows, 16.00 B in 0.007 sec., 272.95 rows/sec., 2.13 KiB/sec.
```
The flow of Pipeline is as follows:
```
+---------------+      +-----------------------+
|MutationSource1| ---> |SerializeDataTransform1|   ------
+---------------+      +-----------------------+         |      +-----------------+      +------------+
|     ...       | ---> |          ...          |   ...   | ---> |MutationTransform| ---> |MutationSink|
+---------------+      +-----------------------+         |      +-----------------+      +------------+
|MutationSourceN| ---> |SerializeDataTransformN|   ------
+---------------+      +-----------------------+
```
`UpdateSource`: check whether need to update and generate a new block and write the object. The process is similar to `DeletionSource`.

```
UPDATE column = expression WHERE condition
```
Replace column to the result of the following expression:
```
if(condition, CAST(expression, type), column)
```

### CODE REVIEW

**Comment by sundy-li** [comment]
This pr may be conflicted a lot with new expression, maybe we can slightly hold on ?

**Comment by BohuTANG** [comment]
> This pr may be conflicted a lot with new expression, maybe we can slightly wait for that merged ?
> 
> I think it's better to create a pr based on expression branch.

OK.
I convert this PR to a draft. 
cc @zhyass @dantengsky

File: src/query/storages/fuse/src/operations/mutation/update/update_source.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+                let col_metas = util::column_metas(&meta_data)?;
+
+                // new block meta.
+                let new_meta = Arc::new(BlockMeta::new(
+                    row_count,
+                    block_size,
+                    file_size,
+                    col_stats,
+                    col_metas,
+                    None,
```
Comment (dantengsky) [inline src/query/storages/fuse/src/operations/mutation/update/update_source.rs:279]: do we need to update the clustering stats here? or there will be a dedicated PR to update the clustering stats?

**Comment by dantengsky** [comment]
@zhyass `DeletionSource` and `UpdateSource` share some common code, it should be better if we could de-duplicate them. it is ok to me to do the refactoring in a dedicated PR.

File: src/query/storages/fuse/src/operations/mutation/update/update_source.rs (hunk 1 above)
Comment (zhyass) [reply src/query/storages/fuse/src/operations/mutation/update/update_source.rs]: > update the clustering stats

https://github.com/datafuselabs/databend/issues/9616

File: src/query/service/tests/it/storages/fuse/operations/mutation/update.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+use databend_query::sessions::QueryContext;
+use databend_query::sessions::TableContext;
+
+use crate::storages::fuse::table_test_fixture::execute_command;
+use crate::storages::fuse::table_test_fixture::execute_query;
+use crate::storages::fuse::table_test_fixture::expects_ok;
+use crate::storages::fuse::table_test_fixture::TestFixture;
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_update_mutator_multiple_empty_segments() -> Result<()> {
```
[the remark this answers is not in the document]
Comment (zhyass) [reply src/query/service/tests/it/storages/fuse/operations/mutation/update.rs]: The ut case has been removed.

File: tests/sqllogictests/suites/base/03_dml/03_0035_update [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+statement ok
+INSERT INTO t VALUES(1, '2022-12-30')
+
+statement ok
+INSERT INTO t VALUES(2, '2023-01-01')
+
+statement ok
+UPDATE t SET a = 3 WHERE b > '2022-12-31'
+
```
[the remark this answers is not in the document]
Comment (zhyass) [reply tests/sqllogictests/suites/base/03_dml/03_0035_update:21]: The test cases have been added.

**Review by BohuTANG — APPROVED** [review]

**Comment by BohuTANG** [comment]
image
👍

**Review by dantengsky — APPROVED** [review]

**Comment by dantengsky** [comment]
@zhyass  just a reminder:

after PR #9460 is merged, i think, the updation process needs to 
- adjust the use of "column identity" 
   not quite sure, seems `bind_update`, and `UpdateInterpreter::execute2` should be adjusted.
- adjust the logic of conflict detection
   the simplest way, by now I can come up with, is just to abort if updating in parallel with a scheme evolution, and the update is committed later.   but apparently, the strategy could be better for some cases (fields being updated not changed, e.g.)

**Comment by BohuTANG** [comment]
@soyeric128 looking forward to add update feature to documents, thanks.
